proposals/AAAA-stdlib-span-properties.md

```diff
@@ -0,0 +1,261 @@
+
+```swift
+struct A: ~Copyable, Escapable {}
+struct B: ~Escapable, Copyable {
+  init(_ a: borrowing A) {}
+}
+extension A {
+  var b: B { B(self) }
+}
+
+func function() {
+    var a = A()
+    var b = a.b // access to `a` begins here
+    read(b)
+    // `b` has ended here, ending access to `a`
+    modify(&a)  // `modify()` can have exclusive access to `a`
+}
+```
+If we were to attempt using `b` again after the call to `modify(&a)`, the compiler would report an overlapping access error, due to attempting to mutate `a` (with `modify(&a)`) while it is already being accessed through `b`'s borrow. Note that the copyability of `B` means that it cannot represent a mutation of `A`; it therefore represents a non-exclusive borrowing relationship.
+
+Given this, we propose to enable the definition of a borrowing relationship via a computed property. With this feature we then propose to add `storage` computed properties to standard library types that can share their internal typed storage, as well as `bytes` computed properties to those standard library types that can safely share their internal storage as untyped memory.
+
+## Detailed Design
+
+A computed property getter of an `Escapable` type returning a non-escapable and copyable type (`~Escapable & Copyable`) establishes a borrowing lifetime relationship of the returned value on the callee's binding. As long as the returned value exists (including local copies,) then the callee's binding is being borrowed. In terms of the law of exclusivity, a borrow is a read-only access. Multiple borrows are allowed to overlap, but cannot overlap with any mutation.
+
+By allowing the language to define lifetime dependencies in this limited way, we can add `Span`-providing properties to standard library types.
+
+#### Extensions to Standard Library types
+
+The standard library and Foundation will provide `storage` and `bytes` computed properties. These computed properties are the safe and composable replacements for the existing `withUnsafeBufferPointer` and `withUnsafeBytes` closure-taking functions.
+
+```swift
+extension Array {
+  /// Share this `Array`'s elements as a `Span`
+  var storage: Span<Element> { get }
+}
+
+extension Array where Element: BitwiseCopyable {
+  /// Share the bytes of this `Array`'s elements as a `RawSpan`
+  var bytes: RawSpan { get }
+}
+
+extension ArraySlice {
+  /// Share this `Array`'s elements as a `Span`
+  var storage: Span<Element> { get }
+}
+
+extension ArraySlice where Element: BitwiseCopyable {
+  /// Share the bytes of this `Array`'s elements as a `RawSpan`
+  var bytes: RawSpan { get }
+}
+
+extension ContiguousArray {
+  /// Share this `Array`'s elements as a `Span`
+  var storage: Span<Element> { get }
+}
+
+extension ContiguousArray where Element: BitwiseCopyable {
+  /// Share the bytes of this `Array`'s elements as a `RawSpan`
+  var bytes: RawSpan { get }
+}
+
+extension String.UTF8View {
```

Comment:
Q: What is the behavior (and what are the performance characteristics) of this operation when a `String` has storage other than contiguous UTF-8? Would it be desirable to return an optional, or alternatively require the user to know or have called `makeContiguousUTF8()` on the `String` as a precondition?

Reply:
Behind the scenes, we are implementing a new "lazy eager bridging" behaviour for bridged `Array` and bridged `String`. They will always succeed, but they will sometimes entail an allocation and copy (i.e. "usually O(1), sometimes O(n)".) An initial step is [here](https://github.com/swiftlang/swift/pull/75148).

Reply:
Cool! The implementation details would of course be out of scope for this text, but might be worth calling out here the performance implications for calling `storage` and `bytes` on these types, pros and cons for usability vs predictability, etc. (given that this is supposed to be a safe, performant API).

Reply:
I appended a paragraph to the detailed design to spell this out.